synapse/handlers/sync.py

```diff
@@ -534,6 +536,10 @@ async def _load_filtered_recents(
 
             prev_batch_token = now_token.copy_and_replace("room_key", room_key)
 
+        if self._msc2716_enabled:
+            # `m.historical` events should not come down /sync
+            recents = filter_historical_events(recents)
```

Comment:
I think this should be up near line 465 for a few reasons:

1. There's an early return on line 473 which would skip this filtering.
2. I think we want to do this before limiting to a certain number of messages.
3. This should be included in the measure block.

What do you think?

Reply:
The problem with [line 464](https://github.com/matrix-org/synapse/blob/80d6dc9783aa80886a133756028984dbf8920168/synapse/handlers/sync.py#L464) is it's before we [handle `loaded_recents` and set as `recents` again](https://github.com/matrix-org/synapse/blob/80d6dc9783aa80886a133756028984dbf8920168/synapse/handlers/sync.py#L502-L523).

So we miss out filtering all of those `loaded_recents` and the Complement test I wrote fails ([`Backfilled events with m.historical do not come down /sync`](https://github.com/matrix-org/complement/pull/68/files))

---

You can see I was playing around with that spot in a previous commit: https://github.com/matrix-org/synapse/pull/9247/commits/7f4c3a645b3a74422b379e4280dc3fd6b15b5a0a#diff-70c84622edff898486bdc02d7a579db4e52a3e0b26c14020bed0ca1fb8ea2018R469


----


➡️  Maybe we should add it to both spots [line 461](https://github.com/matrix-org/synapse/blob/80d6dc9783aa80886a133756028984dbf8920168/synapse/handlers/sync.py#L461) and [line 523](https://github.com/matrix-org/synapse/blob/80d6dc9783aa80886a133756028984dbf8920168/synapse/handlers/sync.py#L523) like we call `filter_events_for_client` in both places.

Reply:
@clokep The diff is "outdated" and the comments are hidden away but still relevant. How does this approach sound?

Reply:
I think it sounds OK. I'll take a look at the updated diff soon.